pkg/dl/dl.go

```diff
@@ -48,35 +49,80 @@ func New(name string, flags int) *DynamicLibrary {
 	}
 }
 
+func useLibDL(action func() error) error {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	return action()
+}
+
+type DLError struct {
+	err string
+}
+
+func (e DLError) Error() string {
+	return e.err
+}
+
+func dlError() error {
+	lastErr := C.dlerror()
+	if lastErr == nil {
+		return nil
+	}
+	return DLError{
+		err: C.GoString(lastErr),
+	}
+}
+
 func (dl *DynamicLibrary) Open() error {
 	name := C.CString(dl.Name)
 	defer C.free(unsafe.Pointer(name))
 
-	handle := C.dlopen(name, C.int(dl.Flags))
-	if handle == C.NULL {
-		return fmt.Errorf("%s", C.GoString(C.dlerror()))
+	var handle unsafe.Pointer
+	if err := useLibDL(func() error {
+		handle = C.dlopen(name, C.int(dl.Flags))
+		if handle == nil {
+			return dlError()
+		}
+		return nil
+	}); err != nil {
+		return err
 	}
+
 	dl.handle = handle
 	return nil
 }
 
 func (dl *DynamicLibrary) Close() error {
-	err := C.dlclose(dl.handle)
-	if err != 0 {
-		return fmt.Errorf("%s", C.GoString(C.dlerror()))
+	if dl.handle == nil {
+		return nil
 	}
+	if err := useLibDL(func() error {
+		if C.dlclose(dl.handle) != 0 {
+			return dlError()
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	dl.handle = nil
 	return nil
 }
 
 func (dl *DynamicLibrary) Lookup(symbol string) error {
 	sym := C.CString(symbol)
 	defer C.free(unsafe.Pointer(sym))
 
-	C.dlerror() // Clear out any previous errors
-	C.dlsym(dl.handle, sym)
-	err := C.dlerror()
-	if unsafe.Pointer(err) == C.NULL {
+	var pointer unsafe.Pointer
```

Comment:
This pointer is local to the check below and could be initialised when calling `C.dlsym` below.

Also, we seem to have lost the call to `C.dlerror()` to clear previous errors. Is this no longer needed / expected?

Reply:
I see no cases where we have to clear `dlerror` since we call it each time error actually happens (and now we know that we won't miss any errors since they will arrive in the same thread as `open`/`close`/`sym` calls). So imo this is no longer needed

Reply:
Returned `dlError()` call.